File: packages/common/functions/messageUpdateIdentity.ts
import { createHash } from 'node:crypto';
import { IUpdateMessage } from '../interfaces/IUpdateMessage';
import { parseSerializedMessageId } from './parseSerializedMessageId';

const MESSAGE_UPDATE_EVENT_ID_VERSION = 'v1';

function nonEmpty(value?: string | null): string | null {
  const normalized = value?.trim();
  return normalized ? normalized : null;
}

function resolveWorkerId(data: IUpdateMessage): string | null {
  return nonEmpty(data.worker_id) ?? nonEmpty(data.data?.worker?.id);
}

export function buildMessageUpdateKafkaKey(data: IUpdateMessage): string {
  const accountId = nonEmpty(data.data?.account?.id) ?? 'unknown-account';
  const workerId = resolveWorkerId(data) ?? 'unknown-worker';
  const messageId =
    nonEmpty(data.data?.message_id) ??
    nonEmpty(data.message?.key?.id) ??
    'unknown-message';
  return `${accountId}:${workerId}:${messageId}`;
}

export function buildMessageUpdateEventId(data: IUpdateMessage): string | null {
  const accountId = nonEmpty(data.data?.account?.id);
  const workerId = resolveWorkerId(data);
  const internalMessageId = nonEmpty(data.data?.message_id);
  const rawProviderMessageId = nonEmpty(data.message?.key?.id);
  const providerMessageId = rawProviderMessageId
    ? (parseSerializedMessageId(rawProviderMessageId)?.stanzaId ??
      rawProviderMessageId)
    : null;

  if (!accountId || !workerId || !internalMessageId || !providerMessageId) {
    return null;
  }

  const canonical = [
    MESSAGE_UPDATE_EVENT_ID_VERSION,
    accountId,
    workerId,
    internalMessageId,
    providerMessageId,
  ].join('\0');

  return `message_update_${MESSAGE_UPDATE_EVENT_ID_VERSION}_${createHash(
    'sha256'
  )
    .update(canonical)
    .digest('hex')}`;
}

export function ensureMessageUpdateIdentity(data: IUpdateMessage): void {
  data.worker_id = resolveWorkerId(data) ?? undefined;
  const existingEventId = nonEmpty(data.event_id);
  if (existingEventId) {
    data.event_id = existingEventId;
    return;
  }

  data.event_id = buildMessageUpdateEventId(data) ?? undefined;
}

File: apps/worker_whatsmeow/internal/app/message_update_identity.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

const messageUpdateEventIDVersion = "v1"

func ensureMessageUpdateEventID(update *UpdateMessage) string {
	if update == nil {
		return ""
	}
	if existing := strings.TrimSpace(update.EventID); existing != "" {
		update.EventID = existing
		return existing
	}
	accountID := strings.TrimSpace(stringValue(update.Data.Account["id"]))
	workerID := strings.TrimSpace(firstNonEmpty(
		update.WorkerID,
		stringValue(update.Data.Worker["id"]),
	))
	internalMessageID := strings.TrimSpace(update.Data.MessageID)
	providerMessageID := canonicalInboundStanzaID(
		valueString(update.Message["key"], "id"),
	)
	if accountID == "" || workerID == "" || internalMessageID == "" || providerMessageID == "" {
		return ""
	}
	canonical := strings.Join([]string{
		messageUpdateEventIDVersion,
		accountID,
		workerID,
		internalMessageID,
		providerMessageID,
	}, "\x00")
	digest := sha256.Sum256([]byte(canonical))
	update.EventID = "message_update_" + messageUpdateEventIDVersion + "_" + hex.EncodeToString(digest[:])
	return update.EventID
}
